File: 43. Magicstar/043.py
from Crypto.Util.number import bytes_to_long as b2l
from Crypto.Util.number import long_to_bytes as l2b

def bs(n):
    # Calc how many zero bytes at the end of n
    pass


def work(n):
    a = [0xff] * (n + 1)
    for _ in range(n):
        num, name = input().split(' ')
        num = int(num)
        name = b2l(l2b(_) + bytes(name[::-1], encoding='ascii'))
        a[num] = name
    m = int(input())
    for _ in range(m):
        q = int(input())
        ans = 0
        for j in range(q):
            x, y = input().split(' ')
            x = int(x)
            y = int(y)
            cmp = a[x] ^ a[y]
            # print(hex(cmp))
            while 0 == (cmp & 0xff):
                cmp >>= 8
                ans += 1
        print(ans)


def work2(n):
    a = [0] * (n + 1)
    for _ in range(n):
        num, name = input().split(' ')
        num = int(num)
        a[num] = name
    m = int(input())
    for _ in range(m):
        q = int(input())
        ans = 0
        for j in range(q):
            x, y = input().split(' ')
            x = int(x)
            y = int(y)
            for k in range(min(len(a[x]), len(a[y]))):
                if a[x][k] != a[y][k]:
                    break
            ans += k
        print(ans)


# -- author: lijw --
if __name__ == '__main__':
    n = int(input())
    work2(n)

File: 43. Magicstar/043.cpp

#include<stdio.h>
#include<string.h>
char name[6001][1001];
char n1[1001],n2[1002];
int n,m,q;
int main(){
	int ans,i,j,k;
	scanf("%d",&n);
	while(n--){
		scanf("%d",&i);
		scanf("%s",&name[i]);
	}
	scanf("%d",&m);
	while(m--){
		scanf("%d",&q);
		ans=0;
		while(q--){
			scanf("%d %d",&i,&j);
			for(k=0;name[i][k]!='\0'&&name[j][k]!='\0';k++){
				if(name[i][k]!=name[j][k])
				break;
			}
			ans+=k;
		}
		printf("%d\n",ans);
	}
	return 0;
} 
